### DISCUSSION TITLE
Slash Commands Without Nvim-Cmp

### DISCUSSION THREAD

**Discussion by collindutter** [olimorris/codecompanion.nvim#381] [General]
Hi! I love the plugin, fantastic work!

I recently switched from [nvim-cmp](https://github.com/hrsh7th/nvim-cmp) to [blink.cmp](https://github.com/Saghen/blink.cmp). Is there any way trigger slash commands without using `nvim-cmp`? And are there any plans to support non-`nvim-cmp` autocompletion providers?

Thanks!

**Comment by olimorris** [comment]
I'm keeping a close on blink. Very excited by what it offers. Docs are still super light and I note that the API is still in "beta quality" but I would love to add support for it when it becomes feasible.

I'd expect a lot of the logic that I've added in for nvim-cmp could just be copied over.

**Reply by collindutter** [reply]
Thanks for the quick reply! Makes sense that you want to allow blink to stabilize before integrating. What about using slash commands without an autocomplete at all? Maybe some keybind/command to execute the slash command under the cursor?

**Reply by GitMurf** [reply]
I am also interested but even more generalized ;) the neovim native completions! They are getting so good in nightly! I now don't use any plugin and just use vim lsp omnifunc and some small tweaks.

This inspired me :) https://boltlessengineer.github.io/posts/neovim-config-without-plugins/#completion

**Reply by olimorris** [reply]
@GitMurf can you show me an example of how your completions look? I would _love_ to move away relying on any completion plugin.

**Reply by GitMurf** [reply]
@olimorris a bunch of info incoming (remember, you asked for it! 😜 ) on how I recommend you setup completions...

But before that real quick, I have been digging through the api and I cannot figure out for the life of me how I can essentially build my own "slash commands picker". It seems all roads lead back to nvim-cmp for initiating slash commands. I just want to build my own simple telescope picker by listing the names of the slash commands and then letting me pick one from there and have it execute as if my Telescope picker was nvim-cmp completion popup. If you can just point me to one that would be wonderful as I can figure out the rest! ... like maybe how to run the `/buffer` slash command so I get the popup to select an open buffer to add as a context reference.

Thanks!

**Reply by GitMurf** [reply]
> @GitMurf can you show me an example of how your completions look? I would _love_ to move away relying on any completion plugin.

Ok, @olimorris here you go :-) My implementation is too messy to be useful for you (been over engineering my own customizations haha) but below is what I used to build mine. You should be able to knock it out in an hour or two. The simplest of it is pretty quick, but of course all the added value items (slippery slope) take the most time haha.

### First off, I mainly used these three sources to build mine from

I especially found useful (and basically mimicked the core of mine from) the first two. The third is of course mini.completions plugin module which I used as a good resource for "ok now how would I accomplish this tweak I want to do" as mini.completion is kind of in the middle of simple "native" config solution VS full blown plugin (cmp).

1. https://github.com/konradmalik/neovim-flake/blob/main/config/native/lua/pde/lsp/completion.lua
2. https://github.com/deathbeam/autocomplete.nvim/blob/main/lua/autocomplete/buffer.lua
3. https://github.com/echasnovski/mini.nvim/blob/main/lua/mini/completion.lua

### Use these GitHub searches to find examples of others using native completions

- https://github.com/search?q=%22v%3Alua.vim.lsp.omnifunc%22+path%3A**%2Fdot*%2F**%2F*.lua&type=code
- https://github.com/search?q=vim.lsp.completion.enable+path%3A*.lua&type=code
- https://github.com/search?q=lsp.completion+%22autotrigger+%3D%22+path%3A*.lua&type=code
- https://github.com/search?q=vim.lsp.completion.trigger+path%3A*.lua&type=code

If you run into any hiccups or have any questions as you try to set it up don't hesitate to reach out! At some point I plan to do an absolute bare minimum implementation with the minimal code needed to get the first LSP completion in your menu, but that is on the back burner right now (as most things are haha).

Good luck!

**Reply by olimorris** [reply]
@GitMurf I can't wait!

So the [cmp file](https://github.com/olimorris/codecompanion.nvim/blob/main/lua/cmp_codecompanion/slash_commands.lua) for slash commands should be the best place.

The `source:complete` method gets all of the slash commands from the config and formats them ready to be displayed in the completion menu.

The `source:execute` method is then what is called when a user selects an item in the menu.

There's a few quirks with slash commands as they can also be from prompts that a user has defined in the prompt library.

**Reply by collindutter** [reply]
Sharing one more native completion resource while the conversation is hot! https://gist.github.com/MariaSolOs/2e44a86f569323c478e5a078d0cf98cc

**Reply by GitMurf** [reply]
> Sharing one more native completion resource while the conversation is hot! https://gist.github.com/MariaSolOs/2e44a86f569323c478e5a078d0cf98cc

Funny I meant to actually share this same resource but evidently I forgot to paste it into my message above 🤣  @MariaSolOs is as good as it gets!!

**Reply by GitMurf** [reply]
> So the [cmp file](https://github.com/olimorris/codecompanion.nvim/blob/main/lua/cmp_codecompanion/slash_commands.lua) for slash commands should be the best place.

So I would have came back and said it was pretty straight forward and not too difficult... but I had thought I was doing something wrong for the longest time. My telescope picker opened fine and showed the buffers, but I selected one and nothing happened. Couldn't figure it out until I accidentally hit tab and noticed it marked it and thought "oh maybe I have to tab select what I want?" 🤦🏻‍♂️🤣 And then it worked.

On a serious note, I am typically used to with telescope that pressing Enter on an item is an "implicit" selection. I imagine this could be something that gets other users too. Not sure if you have a strong opinion one way or the other but I would recommend hitting Enter on an item selects/chooses it.

**Reply by MariaSolOs** [reply]
Aww @GitMurf @collindutter glad that you've found my code snippet helpful! <3

**Reply by olimorris** [reply]
> On a serious note, I am typically used to with telescope that pressing Enter on an item is an "implicit" selection. I imagine this could be something that gets other users too. Not sure if you have a strong opinion one way or the other but I would recommend hitting Enter on an item selects/chooses it.

We don't modify any mappings with Telescope. Whatever you've set in your Telescope setup is carried through to CodeCompanion.

**Reply by GitMurf** [reply]
> We don't modify any mappings with Telescope. Whatever you've set in your Telescope setup is carried through to CodeCompanion.


So is the expectation when you use the "default action" (or whatever they call it), which is Enter key, that should select a buffer for example and add it as a reference in the chat? In pretty much everything in telescope, if you do not Tab select multiple items, pressing Enter the default is to "select / apply" that item. But it seems slash commands are not doing that. 

Let me know if we are not fully understanding each other ;)

**Reply by collindutter** [reply]
So as far as implementing something, I can think of a couple paths:
1. Set a custom [omnifunc](https://neovim.io/doc/user/options.html#'omnifunc') in the CodeCompanion buffer. This would not be _auto_-completed, however, and would require users to trigger via `CTRL-X CTRL-O`.
2. Implement a CodeCompanion language server that provides completions. This feels a bit hacky and overcomplicated, but it would allow the completions to work with any autocompletion provider (native, cmp, blink, etc). Could perhaps draw inspiration from [none-ls](https://github.com/nvimtools/none-ls.nvim/blob/fb7acad9e98ad9c617775451c1aea95157e17942/lua/null-ls/client.lua#L92).

**Reply by GitMurf** [reply]
> 1. Set a custom [omnifunc](https://neovim.io/doc/user/options.html#'omnifunc') in the CodeCompanion buffer. This would not be _auto_-completed, however, and would require users to trigger via `CTRL-X CTRL-O`.

I think this is a great idea and the approach I would take.

@collindutter for the time being, have you done much with omnifunc / creating your own custom sources like this?

I then think it would be fairly trivial for @olimorris to provide an interface to map to, likely based off current Cmp integration, and then could make it generalized instead of needing to support each type of completion plugin / tool / source.

In hacking through the api a bit, it already is fairly easy to get slash commands to work, so I imagine it wouldn't be too difficult (famous last words) 😀

**Reply by collindutter** [reply]
> have you done much with omnifunc / creating your own custom sources like this?

Not at all, I was reading up on all this today 😄 

I wasn't able to find many resources on custom completefuncs, especially in lua. Could maybe use `mini.completion`s [completefunc](https://github.com/echasnovski/mini.completion/blob/7f41cfa486a408c7bdadd4817c5a28888122d739/lua/mini/completion.lua#L363) as a starting point.

**Reply by GitMurf** [reply]
> > We don't modify any mappings with Telescope. Whatever you've set in your Telescope setup is carried through to CodeCompanion.
> 
> So is the expectation when you use the "default action" (or whatever they call it), which is Enter key, that should select a buffer for example and add it as a reference in the chat? In pretty much everything in telescope, if you do not Tab select multiple items, pressing Enter the default is to "select / apply" that item. But it seems slash commands are not doing that.
> 
> Let me know if we are not fully understanding each other ;)

@olimorris beat you to announcing your own fix! 🤣  Thanks for getting on this! 🎉  https://github.com/olimorris/codecompanion.nvim/commit/6184bd4e60d4f301cc5f47416a9a90731eb7f8fb

**Comment by lcmen** [comment]
Thanks for the plugin! It's really awesome. I have some beginner questions.

I also don't use `nvim-cmp`, and I was also wondering if I can use slash command in chat buffer. Any time I type `/buffer`, e.g. `Check current /buffer for code errors` it replies: `Please select the code in the buffer you want me to check`.

> Set a custom [omnifunc](https://neovim.io/doc/user/options.html#'omnifunc') in the CodeCompanion buffer. This would not be auto-completed, however, and would require users to trigger via CTRL-X CTRL-O.

How can I set a custom `omnifunc` for completing codecompanion slash commands for code companion chat buffer?

**Comment by olimorris** [comment]
We have some progress:

https://github.com/user-attachments/assets/1368d784-01e7-4f45-b932-f6ce2f16414f

Ironically, what's been the biggest help is using the `help` slash command and sending over a load of sections from the Neovim manual to Anthropic. I've set a customisable keymap in the config, `<C-/>`, which triggers completion in the chat buffer.

The plan is to then remove cmp as a required dependency and I'll do the same for Telescope soon.

Hopefully ship this tomorrow.

**Reply by collindutter** [reply]
You're a legend, great work @olimorris!

**Comment by olimorris** [comment]
Gentleman, this is live. Please let me know how you get on. The default mapping is `<C-/>` which can be overwritten in the config.

**Reply by collindutter** [reply]
So exciting! I can't seem to get it to trigger though. Note that I rebound to `<C-x>`:
```lua
  opts = {
    strategies = {
      chat = {
        keymaps = {
          completion = {
            modes = {
              i = '<C-x>',
            },
          },
        },
      },
    },
  },
```

https://github.com/user-attachments/assets/8025dab3-0353-4853-8630-87bae7bc1e65

**Reply by olimorris** [reply]
Ah it'll only work if you type `/` and then the key map to complete. Of course you can type anything after the `/` to filter the matches.

Same for `#` and `@`.

Sorry should have made that clear.

**Reply by collindutter** [reply]
Hmm, I think I do that at `:20` in the video. Totally possible there is something messed up in my configuration. I'll keep playing with it, maybe trying with a clean install.
![image](https://github.com/user-attachments/assets/33b73799-b625-4cbf-b783-fb29b20adaac)

**Reply by collindutter** [reply]
Also, with the default mapping, pressing `<C-/>` gives me `^_`. 
![image](https://github.com/user-attachments/assets/650c1e9b-d21b-4e02-8d1a-7cf9e4a7e8a1)

This seems to be a [semi-special](https://vi.stackexchange.com/questions/26611/is-it-possible-to-map-control-forward-slash-with-vim) mapping, maybe even something related to my terminal (kitty).

**Reply by olimorris** [reply]
Eeek yeah it may be. What about if you map it to <c-x><c-o>? The maps are tied to the chat buffer so won't affect any other mappings you have

**Reply by collindutter** [reply]
Seemingly no effect, the `?` is just inserted after `/`. I can make changes to the plugin locally, are there any places that might be good to place some sanity check logs?

**Reply by olimorris** [reply]
Do you have nvim-cmp installed at all? Because it will look to use. Can you do an `:imap <C-\>` in the chat buffer and see if it's set? (Or whatever your mapping is)

**Reply by collindutter** [reply]
I don't have `nvim-cmp` installed and have verified that [this](https://github.com/olimorris/codecompanion.nvim/blob/31c4d0917bf77340fd80f661f38b9ab6f18fd18e/lua/codecompanion/keymaps.lua#L166-L173) returns false.

With no custom mapping, `:imap <C-/>` shows `No mapping found`.  Same if I set a custom one. I have verified the expected mapping appears in the `?` menu. Let me try on a clean nvim.

**Reply by olimorris** [reply]
Spotted it. It wasn't resolving the `has_cmp` conditional properly. Should be working now.

**Reply by collindutter** [reply]
Still having the issue with this minimal. Maybe we should wait to see whether it works for @GitMurf. 
```lua
---@diagnostic disable: missing-fields

--NOTE: Set config path to enable the copilot adapter to work.
--It will search the following paths for the for copilot token:
--  - "$CODECOMPANION_TOKEN_PATH/github-copilot/hosts.json"
--  - "$CODECOMPANION_TOKEN_PATH/github-copilot/apps.json"
vim.env["CODECOMPANION_TOKEN_PATH"] = vim.fn.expand("~/.config")

vim.env.LAZY_STDPATH = ".repro"
load(vim.fn.system("curl -s https://raw.githubusercontent.com/folke/lazy.nvim/main/bootstrap.lua"))()

-- Your CodeCompanion setup
local plugins = {
	{
		"olimorris/codecompanion.nvim",
		dependencies = {
			{ "nvim-treesitter/nvim-treesitter", build = ":TSUpdate" },
			{ "nvim-lua/plenary.nvim" },
			{ "stevearc/dressing.nvim", opts = {} },
			{ "nvim-telescope/telescope.nvim" },
		},
		opts = {
			--Refer to: https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/config.lua
			strategies = {
				--NOTE: Change the adapter as required
				chat = { adapter = "copilot" },
				inline = { adapter = "copilot" },
			},
			opts = {
				log_level = "DEBUG",
			},
		},
	},
}

require("lazy.minit").repro({ spec = plugins })

-- Setup Tree-sitter
local ts_status, treesitter = pcall(require, "nvim-treesitter.configs")
if ts_status then
	treesitter.setup({
		ensure_installed = { "lua", "markdown", "markdown_inline", "yaml" },
		highlight = { enable = true },
	})
end
```

```bash
nvim --clean -u minimal.lua
```

**Reply by olimorris** [reply]
@collindutter I've just pushed an updated commit. I wasn't properly evaluating if cmp existed which meant it wasn't setting the keymap.

**Reply by collindutter** [reply]
That did it! 🎉 
![image](https://github.com/user-attachments/assets/3f42169f-2e74-4ce3-8022-7710862783ef)

I still had to rebind to `<C-_>` as per [this thread](https://vi.stackexchange.com/questions/26611/is-it-possible-to-map-control-forward-slash-with-vim), not sure if that's a better default or not.

EDIT: fixed link.

**Reply by olimorris** [reply]
I won't take that chance. I'll make `<c-_>` the default. Thanks for testing.

**Reply by collindutter** [reply]
Thanks for the awesome plugin and responsiveness! Going to mark this as resolved.

**Reply by GitMurf** [reply]
Traveling this weekend but will try when back home Sunday!

**Comment by rpbaptist** [comment]
Is there any way this can work with FzfLua instead of Telescope?

**Reply by olimorris** [reply]
Funny you mention it. I'm refactoring the slash commands tomorrow and will be adding fzf-lua for commands alongside Telescope.

I believe I have a "default" option for everything except help tags.

**Reply by olimorris** [reply]
@rpbaptist fzf_lua is an option right now for everything but the help tags FYI

**Comment by olimorris** [comment]
I'm working on the `blink.cmp` integration. There's an outstanding [issue](https://github.com/Saghen/blink.cmp/issues/440) which needs to be resolved on their side but should be straightforward.

**Reply by GitMurf** [reply]
How is blink? I haven't tried it as I am happy with my simple custom completions implementation, but I keep hearing everyone talk about it!

**Reply by olimorris** [reply]
I only installed it yesterday and my comparison point is nvim-cmp, so too early to really call it. The API is very clean though, install was a breeze. The pull of frecency and auto bracket support won me over. And, I dug through the source code for about 10 mins yesterday and was able to workout how I could add support for it in CodeCompanion. So on that basis, it's a big thumbs up.

I liken it to when Lazy.nvim came along after Packer. You had a decent repo to learn from and could also avoid the big architectural mistakes which hindered new features (I seem to recall versioning and rollbacks being a real challenge in Packer). And in the case of blink.cmp, you can add ridiculous speed (not that it was ever an issue for me) with sensible result algorithms.

**Comment by olimorris** [comment]
Happy to report that `v10.5.0` supports [blink.cmp](https://github.com/Saghen/blink.cmp) following the resolution of [#440](https://github.com/Saghen/blink.cmp/issues/440).

Follow the instructions in the README in order to set it up.

**Reply by collindutter** [reply]
Woohoo! Just confirmed that it is working 
![image](https://github.com/user-attachments/assets/60174c0e-0b1c-4576-b69a-0fe6146c2bff)

Worth nothing that I had to install blink from HEAD since that patch is not in a release yet. And I _think_ we can use `opts_extend` to avoid redeclaring the providers, but I can't seem to get that part working..

```lua
return {
  'saghen/blink.cmp',
  build = 'cargo build --release',
  -- version = 'v0.*',
  enabled = true,
  opts_extend = { 'sources.completion.enabled_providers' },
  opts = {
    sources = {
      completion = {
        enabled_providers = { 'codecompanion' },
      },
      providers = {
        codecompanion = {
          name = 'CodeCompanion',
          module = 'codecompanion.providers.completion.blink',
          enabled = true,
        },
      },
    },
  },
}
```
